### PR TITLE
[Merged by Bors] - refactor(FieldTheory/NormalClosure): change definition and add API

### PR DESCRIPTION
**Pull request by tb65536**
This PR adds API and changes the definition of the normal closure of $F\leq K\leq L$ to be an intermediate field of L/F, rather than an intermediate field of L/K. For example, I think it would be more common to say that the normal closure of $\mathbb{Q}(\sqrt[3]{2})/\mathbb{Q}$ is $\mathbb{Q}(\sqrt[3]{2},\zeta_3)/\mathbb{Q}$ rather than $\mathbb{Q}(\sqrt[3]{2},\zeta_3)/\mathbb{Q}(\sqrt[3]{2})$. This change also means that the normal closure goes from being a dependent function `(K : Type) → IntermediateField K L` to being a non-dependent function `Type → IntermediateField F L`, making it easier to compare across the Galois corespondence.

Supersedes https://github.com/leanprover-community/mathlib/pull/18971

---

- [x] depends on: #6273
- [x] depends on: #6289

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by riccardobrasca** [comment]
I am not sure I understand the point of the new definition, since as you said [normalClosure](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/Normal.html#normalClosure) is already an `IntermediateField`.

**Comment by tb65536** [comment]
> I am not sure I understand the point of the new definition, since as you said [normalClosure](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/Normal.html#normalClosure) is already an `IntermediateField`.

Yes sorry, I got myself confused. The difference is that if you apply the old definition to an intermediate field, then you have to restrict scalars to get back to the same intermediate field type. It's more convenient in this case to have a function `K : IntermediateField F L → normalClosure K : IntermediateField F L`, rather than a dependent function `K : IntermediateField F L → normalClosure K : IntermediateField K L`. This is the same sort of situation as with `IntermediateField.adjoin`.

**Comment by tb65536** [comment]
I especially have in mind the Galois correspondence, where `normalClosure : IntermediateField F L → IntermediateField F L` corresponds to `normalCore : Subgroup G → Subgroup G`.

**Comment by riccardobrasca** [comment]
> I especially have in mind the Galois correspondence, where `normalClosure : IntermediateField F L → IntermediateField F L` corresponds to `normalCore : Subgroup G → Subgroup G`.

Ah, so the point is if `normalClosure K` should be an `IntermediateField K L` (as the current definition) or an `IntermediateField F L`. I agree that your definition seems more reasonable, but I am not 100% sure. But having both seems a little strange.

**Comment by tb65536** [comment]
Yeah, having both does seems strange. I'll look into refactoring the current definition to be `IntermediateField F L`, and see how it feels.

**Comment by riccardobrasca** [comment]
One problem with your definition is that `normalClosure K` is not a `K`-algebra automatically, so in some sense you loose information.

**Comment by tb65536** [comment]
The refactor seemed to go through smoothly, and I've added back the lost instances.

**Comment by riccardobrasca** [comment]
It looks like a reasonable modification to me, but I would like to have a second opinion. Maybe you can ask on Zulip.

File: Mathlib/FieldTheory/Normal.lean [hunk 1]
```diff
@@ -474,89 +474,3 @@ theorem isSolvable_of_isScalarTower [Normal F K₁] [h1 : IsSolvable (K₁ ≃
 #align is_solvable_of_is_scalar_tower isSolvable_of_isScalarTower
 
 end lift
-
-section normalClosure
```
Comment (eric-wieser) [inline Mathlib/FieldTheory/Normal.lean:478]: Can you make a quick PR that moves this content with no other changes, to make the diff easier to folllow?

Comment (tb65536) [reply Mathlib/FieldTheory/Normal.lean:478]: Sure, will do.

File: Mathlib/FieldTheory/NormalClosure.lean [hunk 2]
```diff
@@ -22,78 +22,131 @@ variable (F K L : Type _) [Field F] [Field K] [Field L] [Algebra F K] [Algebra F
   [IsScalarTower F K L]
 
 /-- The normal closure of `K` in `L`. -/
-noncomputable def normalClosure : IntermediateField K L :=
```
Comment (eric-wieser) [inline Mathlib/FieldTheory/NormalClosure.lean:25]: What's the replacement if I want `IntermediateField K L` after all?

Comment (tb65536) [reply Mathlib/FieldTheory/NormalClosure.lean:25]: You can get it from the `IsScalarTower` instance (turn the `IsScalarTower` into an `algHom`, and take the `fieldRange`).

Comment (eric-wieser) [reply Mathlib/FieldTheory/NormalClosure.lean:25]: Can you add a lemma that states that `(construction_you_describe).restrictScalars K = normalClosure` (or if it's already there, can you tell me its name)?

Comment (tb65536) [reply Mathlib/FieldTheory/NormalClosure.lean:25]: Done: https://github.com/leanprover-community/mathlib4/pull/6163/commits/1ac7f5eb3905a3568712aecb341d03fe5bdd715c

**Comment by eric-wieser** [comment]
I think the PR title and description are out of date: the file is no longer new, and at a glance nor is `IntermediateField.normalClosure`!

File: Mathlib/FieldTheory/NormalClosure.lean [hunk 3]
```diff
@@ -73,12 +82,74 @@ instance is_finiteDimensional [FiniteDimensional F K] :
...
-#align normal_closure.is_finite_dimensional normalClosure.is_finiteDimensional
 
-instance isScalarTower : IsScalarTower F (normalClosure F K L) L :=
-  -- Porting note: the last argument here `(⨆ (f : K →ₐ[F] L), f.fieldRange).toSubalgebra`
-  -- was just written as `_` in mathlib3.
-  IsScalarTower.subalgebra' F L L (⨆ (f : K →ₐ[F] L), f.fieldRange).toSubalgebra
-#align normal_closure.is_scalar_tower normalClosure.isScalarTower
+noncomputable instance algebra : Algebra K (normalClosure F K L) :=
+  ((inclusion (toAlgHom F K L).fieldRange_le_normalClosure).comp
+    (toAlgHom F K L).rangeRestrict).toAlgebra
```
Comment (eric-wieser) [inline Mathlib/FieldTheory/NormalClosure.lean:88]: This almost certainly creates a diamond that didn't exist before: `RingHom.toAlgebra` is almost always a bad idea.
You can fix it with
```suggestion
noncomputable instance algebra : Algebra K (normalClosure F K L) :=
  IntermediateField.algebra <|
    { (⨆ f : K →ₐ[F] L, f.fieldRange).toSubfield with
      -- Porting note: could not inherit neg_mem
      neg_mem' := fun _ hx => Subfield.neg_mem (⨆ f : K →ₐ[F] L, f.fieldRange).toSubfield hx
      algebraMap_mem' := fun r =>
        le_iSup (fun f : K →ₐ[F] L => f.fieldRange) (IsScalarTower.toAlgHom F K L) ⟨r, rfl⟩ }
```
but this seems to defeat the point of the refactor.

Comment (tb65536) [reply Mathlib/FieldTheory/NormalClosure.lean:88]: Well, the point of the refactor is to make `normalClosure` easier to use, so it's not a big deal that the algebra instance retains the messiness of the original definition.

Comment (tb65536) [reply Mathlib/FieldTheory/NormalClosure.lean]: I pushed your definition (slightly golfed)

File: Mathlib/FieldTheory/NormalClosure.lean [hunk 4]
```diff
@@ -71,12 +82,75 @@ instance is_finiteDimensional [FiniteDimensional F K] :
...
+
+lemma normal_iff_forall_fieldRange_eq : Normal F K ↔ ∀ σ : K →ₐ[F] L, σ.fieldRange = K :=
+⟨@AlgHom.fieldRange_of_normal F L _ _ _ K, normal_iff_forall_fieldRange_le.2 ∘ fun h σ ↦ (h σ).le⟩
+
+lemma normal_iff_forall_map_eq : Normal F K ↔ ∀ σ : L →ₐ[F] L, K.map σ = K := by
+  refine' ⟨fun h σ ↦ K.fieldRange_val ▸ _, fun h ↦ normal_iff_forall_map_le.2 (fun σ ↦ (h σ).le)⟩
+  rw [AlgHom.map_fieldRange, normal_iff_forall_fieldRange_eq.1 h, fieldRange_val]
+
+lemma normal_iff_forall_map_eq' : Normal F K ↔ ∀ σ : L ≃ₐ[F] L, K.map ↑σ = K :=
+⟨fun h σ ↦ normal_iff_forall_map_eq.1 h σ, fun h ↦ normal_iff_forall_map_le'.2 (fun σ ↦ (h σ).le)⟩
```
Comment (jcommelin) [inline Mathlib/FieldTheory/NormalClosure.lean:154]: Would it make sense to make these into a TFAE?

**Comment by tb65536** [comment]
> This comment is a bit misleading, the `normalClosure` does not have the signature that you claim it does in the PR description; it is actually `(K : Type) -> (IntermediateField K L)`.

Good point. I've updated the PR description.

**Comment by eric-wieser** [comment]
I pushed another edit, your spelling looked like it was describing `k : K`.

File: Mathlib/FieldTheory/NormalClosure.lean (hunk 4 above)
Comment (tb65536) [reply Mathlib/FieldTheory/NormalClosure.lean]: It's a little longer, but looks cleaner: https://github.com/leanprover-community/mathlib4/pull/6163/commits/0471830b4d449b984bc84a01c8b42d1875ff7c62

**Comment by eric-wieser** [comment]
@jcommelin, I'd argue that a TFAE is worse because:

* In this particular case the overall proof is longer
* It's harder to use, because you have to remember the indices of the statement you want
* It's harder to find, because you can't just guess the name of the sub-statements any more
* It's totally reasonable to expect someone to navigate the API by going via the "hub and spoke" model where `Normal` is the canonical spelling and all the other are consequences. In my opinion, TFAEs are most useful where there is no canonical spelling (or as an implementation detail to reduce the proof burden when building the API).

This came up when definig `Algebra.ModEq`, where we realized that a canonical spelling was more useful than a large TFAE.

**Comment by kbuzzard** [comment]
This looks great to me. Can you show it's a `ClosureOperator`?

**Comment by eric-wieser** [comment]
I think @YaelDillies concluded that ClosureOperator was poorly designed and not worth trying to use; I think a galois connection would be more sensible?

**Comment by tb65536** [comment]
Wouldn't defining a Galois connection require defining a type of normal intermediate fields? Is that something that we want to do?

File: Mathlib/FieldTheory/NormalClosure.lean [hunk 5]
```diff
@@ -71,12 +83,88 @@ instance is_finiteDimensional [FiniteDimensional F K] :
...
+  · exact le_iSup_of_le f le_rfl
+
+lemma normalClosure_mono (h : K ≤ K') : normalClosure F K L ≤ normalClosure F K' L := by
+  rw [normalClosure_def', normalClosure_def']
+  exact iSup_mono (fun f ↦ map_mono f h)
+
+variable (F L)
+
+/-- `normalClosure` as a `ClosureOperator`. -/
+noncomputable def closureOperator : ClosureOperator (IntermediateField F L) where
```
Comment (eric-wieser) [inline Mathlib/FieldTheory/NormalClosure.lean:137]:
```suggestion
@[simps]
noncomputable def closureOperator : ClosureOperator (IntermediateField F L) where
```

**Comment by YaelDillies** [comment]
You can try using `ClosureOperator` because it's the right situation in theory: You have a Galois correspondence between an order and a suborder of it, and you don't want to define that suborder as a separate type. But don't expect too much of it, as Eric said I dislike my own API here.

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**